stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,354 @@
+  Optional<Mapping> stringTable;
+
+public:
+  StaticBinaryELF() {
+    getExecutablePathName();
+
+    programHeaders = reinterpret_cast<const Elf_Phdr *>(getauxval(AT_PHDR));
+    if (programHeaders == nullptr) {
+      return;
+    }
+    // If an interpreter is set in the program headers then this is a
+    // dynamic executable and therefore not valid.
+    for (size_t idx = 0; idx < elfHeader.e_phnum; idx++) {
+      if (programHeaders[idx].p_type == PT_INTERP) {
+        programHeaders = nullptr;
+        return;
+      }
+    }
+
+    if (!fullPathName.empty()) {
+      mmapExecutable();
+    }
+  }
+
+  const char *getPathName() {
+    return fullPathName.empty() ? nullptr : fullPathName.c_str();
+  }
+
+  void *getSectionLoadAddress(const void *addr) {
+    if (programHeaders) {
+      auto searchAddr = reinterpret_cast<Elf_Addr>(addr);
+
+      for (size_t idx = 0; idx < elfHeader.e_phnum; idx++) {
+        auto header = &programHeaders[idx];
+        if (header->p_type == PT_LOAD && searchAddr >= header->p_vaddr
+            && searchAddr <= (header->p_vaddr + header->p_memsz)) {
+          return reinterpret_cast<void *>(header->p_vaddr);
+        }
+      }
+    }
+    return nullptr;
+  }
+
+  // Lookup a function symbol by address.
+  const Elf_Sym *findSymbol(const void *addr) {
+    if (symbolTable.hasValue()) {
+      auto searchAddr = reinterpret_cast<Elf_Addr>(addr);
+      auto symbols = symbolTable->data<Elf_Sym>();
+      const Elf_Sym *bestMatch = nullptr;
+      unsigned long bestDistance = ULONG_MAX;
+
+      for (size_t idx = 0; idx < symbols.size(); idx++) {
+        auto symbol = &symbols[idx];
+        if (ELF_ST_TYPE(symbol->st_info) == STT_FUNC
+            && searchAddr >= symbol->st_value) {
+
+          auto tmpDistance = searchAddr - symbol->st_value;
+          if (tmpDistance < symbol->st_size) {
+            return symbol;
+          }
+          // The searchAddress is past the end of this symbol's region, keep
+          // track of which symbol end address the searchAddress is closest to.
+          tmpDistance -= symbol->st_size;
+          if (tmpDistance < bestDistance) {
+            bestMatch = symbol;
+            tmpDistance = bestDistance;
+          }
+        }
+      }
+      return bestMatch;
+    }
+    return nullptr;
+  }
+
+  const char *symbolName(const Elf_Sym *symbol) {
+    if (stringTable.hasValue()) {
+      auto strings = stringTable->data<char>();
+      if (symbol->st_name < strings.size()) {
+        return &strings[symbol->st_name];
+      }
+    }
+    return nullptr;
+  }
+
+private:
+  // This is Linux specific - find the full path of the executable
+  // by looking in /proc/self/maps for a mapping holding the current
+  // address space. For a static binary it should only be mapping one
+  // file anyway. Don't use /proc/self/exe as the symlink will be removed
+  // if the main thread terminates - see proc(5).
+  void getExecutablePathName() {
```

Comment:
Instead of making this entire file linux specific, make this method linux specific and just `#error` on other targets.  This remainder of the code should apply to the BSDs and Haiku.

Reply:
Done.